Refactor/search cache

skaik-mo opened the pull request:
Refactored search keyword cleanup with a new Worker + UseCase (runs every 1 hour) and removed unused classes/fields.
Renamed explore module to search with updated packages, classes, and endpoints.

skaik-mo committed:
Refactor: Remove unused search keyword cleaner and use case

This commit removes the `KeywordsCacheCleaner` and `KeywordsCacheCleanerImp` classes, as their functionality is now handled by the `ClearExpiredSearchHistoryUseCase`.

The `SearchCacheCleanupWorker` is introduced to periodically clear expired search history using the new use case. This worker is scheduled to run every 15 minutes.

The `SearchRepository` and `LocalSearchDataSource` have been updated to include a `clearExpiredSearch` method. The `SearchKeywordDao` now has a corresponding `clearExpiredSearch` query that deletes search keywords older than 15 minutes.

The `SearchKeywordCacheDto` no longer stores `moviesPages`, `seriesPages`, and `actorsPages` as these were unused.

The `LocalDataSourceBindModule` no longer binds `KeywordsCacheCleaner`.

Tests for `ClearExpiredSearchHistoryUseCase` have been added.

skaik-mo committed:
Refactor: Rename explore module to search

This commit renames the `explore` module and its associated classes, interfaces, and packages to `search`.

The following changes were made:
- Renamed `com.giraffe.media.explore` package to `com.giraffe.media.search`.
- Renamed classes and interfaces within the module to reflect the new name (e.g., `ExploreViewModel` to `SearchViewModel`).
- Updated import statements across the codebase to use the new package name.
- Renamed test files and dummy data to align with the new module name.
- Updated dependency injection modules to reflect the new package and class names.
- Deleted `SearchApiServiceRetrofit.kt` from the old `explore` package and created a new one in the `search` package with updated endpoint.

skaik-mo committed:
Reduce SearchCacheCleanupWorker frequency

The worker will now run every 1 hour instead of every 15 minutes.

AhmedMohamed056 commented on app/src/main/java/com/giraffe/cineverseapp/CineVerseApp.kt:
```diff
@@ -38,4 +40,15 @@ class CineVerseApp : Application(), Configuration.Provider {
+    private fun setupSearchCacheCleanupWorker() {
+        val workRequest = PeriodicWorkRequestBuilder<SearchCacheCleanupWorker>(
+            1, TimeUnit.HOURS
+        ).build()
```
This is one hour? I think the task one day

skaik-mo replied:
This `SearchCacheCleanupWorker` is specifically for search history, not the data displayed in the Home screen.

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+
+    override suspend fun getSearchKeywords(query: String): Flow<List<SearchKeyword>> {
+        return safeCall {
+            if (query.isBlank()) {
```
it's return flow you need safeFlow not safeCall

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private suspend fun getSearchHistoryWhenHasQuery(query: String): Flow<List<SearchKeyword>> {
+        val history = getLocalSearchKeywords(query)
```
and here

skaik-mo replied:
We don’t need `safeFlow` here since it’s already being called inside `getSearchKeywords`.

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private fun getSearchHistoryWhenQueryIsBlank(): Flow<List<SearchKeyword>> {
+        return local.getSearchHistory()
```
and here

skaik-mo replied:
We don’t need `safeFlow` here since it’s already being called inside `getSearchKeywords`.

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private fun getLocalSearchKeywords(query: String): Flow<List<SearchKeyword>> {
+        return local.getSearchKeywords(query)
```
and here

skaik-mo replied:
We don’t need `safeFlow` here since it’s already being called inside `getSearchKeywords`.

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,74 @@
+    }
+
+    private suspend fun getRemoteSearchKeywords(query: String): List<SearchKeyword> {
+        return remote.getSearchKeywords(query)
```
this and otheres safeCall

skaik-mo replied:
We don’t need `safeFlow` here since it’s already being called inside `getSearchKeywords`.

dev-bilal-azzam replied:
No, this actually needs safeCall because the safeFlow won't catch the exceptions thrown by this rmote call

mohamedebrahem13 commented on app/src/main/java/com/giraffe/cineverseapp/worker/SearchCacheCleanupWorker.kt:
```diff
@@ -0,0 +1,25 @@
+) : CoroutineWorker(appContext, workerParams) {
+    override suspend fun doWork(): Result {
+        return try {
+            clearExpiredSearchHistoryUseCase()
```
you can make the use case it self return the success not the worker also it will go in throw infinity loop when it failed

poula22 replied:
@mohamedebrahem13 I think usecase shoudn't know anything about the Result class, but it can return Boolean to know if it's cleared or not then apply Result.success() or Result.error()

mohamedebrahem13 replied:
ده اللي اققصده بس تقريبا فهمتني غلط

skaik-mo committed:
Refactor: Improve error handling in SearchRepositoryImpl

- Replace `safeCall` with `safeFlow` for better Flow integration and error handling.
- Modify `getSearchHistoryWhenHasQuery` to handle local search keywords as a Flow, improving consistency.
- Make `safeFlow`'s `block` parameter a suspend function for greater flexibility.

skaik-mo committed:
Feat: Limit retries for search cache cleanup worker

The `SearchCacheCleanupWorker` will now retry a maximum of 3 times in case of failure. If it still fails after 3 attempts, it will mark the work as failed.

skaik-mo committed:
Merge branch 'develop' into refactor/search-cache

* develop: (22 commits)
  remove launch effect from textField
  resolve choose image don't work correctly
  Keep text field value and cursor position in sync with external changes
  Refactor: Rename MovieLocalDataSourceImp to MovieLocalDataSourceImpl
  Navigate to search screen on recording icon click
  Add navigateToSearchScreen to SearchResultRoute
  update web View to show dialog add image and delete image
  restricted web view side urls
  fix lazy items keys
  remove debug code
  load movie and series details after adding rate
  fix session id encryption
  add key to the lazy layouts
  remove unknown genre keyword from card result
  support pagination for show more screens
  Clear recently viewed movies on logout
  Refactor: Standardize network constant names
  Refactor: Reorganize MovieRepositoryImpl methods
  Refactor: Consolidate movie caching logic in LocalDataSource
  Refactor: Standardize movie cache clearing and deletion methods
  ...

# Conflicts:
#	app/src/main/java/com/giraffe/cineverseapp/di/localDataSourceModule.kt

mohannadahmed00 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+    }
+
+    private suspend fun getRemoteSearchKeywords(query: String): List<SearchKeyword> {
+        return remote.getSearchKeywords(query)
```
this function need safeCall because it returns just list not flow so the safeFlow of parent function can't handle its exception

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+
+    override suspend fun deleteSearchKeyword(keyword: String) =
+        local.deleteSearchKeyword(keyword)
+
```
safeCall here

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+
+    override suspend fun clearSearchHistory() =
+        local.clearSearchHistory()
+
```
here

AhmedMohamed056 commented on repository/media/src/main/java/com/giraffe/media/search/SearchRepositoryImpl.kt:
```diff
@@ -0,0 +1,75 @@
+
+    override suspend fun clearExpiredSearch() =
+        local.clearExpiredSearch()
+
```
and here

skaik-mo committed:
Merge branch 'develop' into refactor/search-cache

* develop:
  change use case for filter discover to take list of int genres
  resolve config from develop
  Refactor: Implement retry limit for CacheCleanupWorker
  fix recently viewed movies pagination sorting
  fix recently viewed movies pagination
  refactor name of function in discover movies and series from remote
  fix recently viewed series pagination
  feature(remote): extend series fetching to support multiple filters (genres, keywords, minVote)
  feature(remote): extend movies fetching to support multiple filters (genres, keywords, minVote)

skaik-mo committed:
Refactor: Wrap SearchRepositoryImpl local calls in safeCall

The local data source calls in `SearchRepositoryImpl` for deleting search keywords, clearing search history, and clearing expired searches are now wrapped in `safeCall` for improved error handling.

skaik-mo committed:
Refactor: Update recently viewed media use cases to support pagination

This commit modifies the `GetRecentlyViewedSeriesUseCase` and `GetRecentlyViewedMoviesUseCase` to accept `page` and `pageSize` parameters, enabling pagination for recently viewed media.

The corresponding repository and data source layers have also been updated to handle these new parameters.

Tests for these components have been adjusted to reflect the changes.

skaik-mo committed:
Merge branch 'develop' into refactor/search-cache

* develop: (23 commits)
  disable play button if youtube video id is blank
  refactor: test cases in media domain
  Refactor: Remove ScreenStates composable from SettingsScreen
  Refactor: Standardize MovieCacheDto field names
  Format release date for media items
  Refactor: Rename extension function orEmpty to orZero for Floats
  fix: ci bug
  fix: ci bug
  fix: ci bug
  fix: ci bug
  fix: ci bug
  use gradle dsl
  fix: uint test error appeared in CI
  refactor: replace hard coded plugin "java-library" with `java-library`
  fix: ci bug
  fix: ci bug
  feature: use custom gradle plugin in repository modules :repository:media and :repository:user
  refactor: -remove necessary comments in :presentation:details
  fix gradle coverage
  refactor: move firebase dependency to firebase bundle
  ...

# Conflicts:
#	datasource/remote/explore/build.gradle.kts
#	domain/media/src/test/kotlin/com/giraffe/media/explore/dummydata/DummyData.kt
#	domain/media/src/test/kotlin/com/giraffe/media/explore/util/DummyData.kt
#	domain/media/src/test/kotlin/com/giraffe/media/movie/usecase/GetRecentlyViewedMoviesUseCaseTest.kt
#	domain/media/src/test/kotlin/com/giraffe/media/search/dummydata/DummyData.kt
#	domain/media/src/test/kotlin/com/giraffe/media/search/usecase/DeleteSearchKeywordUseCaseTest.kt
#	domain/media/src/test/kotlin/com/giraffe/media/search/usecase/GetSearchKeywordsUseCaseTest.kt
#	domain/media/src/test/kotlin/com/giraffe/media/series/usecase/GetRecentlyViewedSeriesUseCaseTest.kt

skaik-mo committed:
Merge branch 'develop' into refactor/search-cache

* develop:
  catch exception in encryption service
  fix encoding session id

skaik-mo committed:
Refactor: Simplify safeFlow implementation

The `safeFlow` function has been refactored to directly apply the `catch` operator to the flow returned by the `block` function, removing the unnecessary `flow { emitAll(block()) }` wrapper.

Additionally, the `getSearchHistoryWhenHasQuery` function in `SearchRepositoryImpl.kt` no longer needs to be a suspend function as it now returns a `Flow` directly.

skaik-mo committed:
Merge branch 'develop' into refactor/search-cache

* develop: (43 commits)
  fix safe collect in profile
  add test for Get Movies By Genre Ids Use Case
  fix issues in based on true events and cinematic masterpieces
  fix base view model safe collect
  refactor collection list section ui on click
  Update domain/media/src/test/kotlin/com/giraffe/media/collections/usecase/ClearCollectionsCacheUseCaseTest.kt
  add six strategy to featured collection
  fix test coverage verification
  handle clear collections after logout
  create use case to get movies and series by sort
  create use case to get movies by keywords id
  create use case to get series by keywords id
  delete get filtered use case
  delete discover series function in repo, and add getSeriesByKeywordsId, getSeriesBySort
  delete get filtered use case
  delete discover movies function in repo, and add getMoviesByKeywordsId, getMoviesBySort
  handle clear collections after logout
  refactor late night strategy
  handle getting collections as flow
  - create use case to get series by list of genres - create use case to get series by list of genres - remove use case filter series - remove use case filter movies
  ...

# Conflicts:
#	app/src/main/java/com/giraffe/cineverseapp/data/database/CineVerseDatabase.kt
#	app/src/main/java/com/giraffe/cineverseapp/di/databaseModule.kt
#	app/src/main/java/com/giraffe/cineverseapp/di/localDataSourceModule.kt
#	repository/media/src/test/java/com/giraffe/media/series/SeriesRepositoryImplTest.kt
